File: frontend/src/lib/recipeFlowLayout.ts
import type {
  BackendMethodAction,
  BackendMethodDocument,
  BackendMethodTableDocument,
} from '../api/client'

export interface FlowRow {
  id: string
  kind: 'ingredient' | 'setup'
  bindingId?: string
  ingredientLineageId?: string
  actionId?: string
  label: string
}

export interface FlowColumn {
  actionId: string
  stageId: string
  stageTitle: string
  index: number
  label: string
}

export interface FlowCell {
  actionId: string
  column: number
  rowStart: number
  rowSpan: number
  inputBindingIds: string[]
  predecessorActionIds: string[]
  kind: 'operation' | 'setup' | 'merge'
}

export interface FlowConnector {
  edgeId: string
  kind: 'sequence' | 'merge' | 'fork'
  fromActionId: string
  toActionId: string
  fromRow: number
  toRow: number
  column: number
}

export interface FlowLane {
  stageId: string
  title: string
  startColumn: number
  endColumn: number
}

export interface FlowRenderWarning {
  code: string
  message: string
  blocking: boolean
  entityId?: string
}

export interface FlowTableLayout {
  rows: FlowRow[]
  columns: FlowColumn[]
  cells: FlowCell[]
  connectors: FlowConnector[]
  lanes: FlowLane[]
  warnings: FlowRenderWarning[]
}

const threshold = .65

function actionSortKey(action: BackendMethodAction) {
  return `${String(action.position).padStart(8, '0')}:${action.id}`
}

function overlap(left: { rowStart: number; rowSpan: number }, right: { rowStart: number; rowSpan: number }) {
  return left.rowStart < right.rowStart + right.rowSpan && right.rowStart < left.rowStart + left.rowSpan
}

function topologicalActions(method: BackendMethodDocument, warnings: FlowRenderWarning[]) {
  const actions = new Map(method.actions.map(action => [action.id, action]))
  const incoming = new Map<string, string[]>(method.actions.map(action => [action.id, []]))
  const outgoing = new Map<string, string[]>(method.actions.map(action => [action.id, []]))
  for (const edge of method.edges) {
    if (!actions.has(edge.from_action_id) || !actions.has(edge.to_action_id)) {
      warnings.push({ code: 'dangling_edge', message: 'A flow connector points to an operation that no longer exists.', blocking: true, entityId: edge.id })
      continue
    }
    incoming.get(edge.to_action_id)!.push(edge.from_action_id)
    outgoing.get(edge.from_action_id)!.push(edge.to_action_id)
  }
  const indegree = new Map([...incoming].map(([id, parents]) => [id, parents.length]))
  const queue = method.actions.filter(action => indegree.get(action.id) === 0).sort((a, b) => actionSortKey(a).localeCompare(actionSortKey(b)))
  const order: string[] = []
  while (queue.length) {
    const next = queue.shift()!
    order.push(next.id)
    for (const childId of [...(outgoing.get(next.id) ?? [])].sort((left, right) => actionSortKey(actions.get(left)!).localeCompare(actionSortKey(actions.get(right)!)))) {
      const nextDegree = (indegree.get(childId) ?? 0) - 1
      indegree.set(childId, nextDegree)
      if (nextDegree === 0) {
        queue.push(actions.get(childId)!)
        queue.sort((a, b) => actionSortKey(a).localeCompare(actionSortKey(b)))
      }
    }
  }
  if (order.length !== method.actions.length) {
    warnings.push({ code: 'graph_cycle', message: 'This flow contains a cycle; connectors are shown in source order until it is repaired.', blocking: true })
    for (const action of [...method.actions].sort((a, b) => actionSortKey(a).localeCompare(actionSortKey(b)))) {
      if (!order.includes(action.id)) order.push(action.id)
    }
  }
  return { order, incoming, outgoing, actions }
}

export function createFlowTableLayout(method: BackendMethodDocument, table?: BackendMethodTableDocument): FlowTableLayout {
  const savedTable = table ?? {
    schema_version: 1 as const,
    labels: [],
    row_order: [],
    column_hints: [],
    setup_action_ids: [],
    terminal_action_ids: [],
    omissions: [],
  }
  const warnings: FlowRenderWarning[] = []
  const { order, incoming, outgoing, actions } = topologicalActions(method, warnings)
  const inputBindings = method.ingredient_bindings.filter(binding => binding.role !== 'reference')
  const bindingById = new Map(inputBindings.map(binding => [binding.id, binding]))
  const rows: FlowRow[] = []
  const setupIds = new Set(savedTable.setup_action_ids.filter(actionId => actions.has(actionId)))
  for (const actionId of order) {
    if (setupIds.has(actionId)) rows.push({ id: `setup:${actionId}`, kind: 'setup', actionId, label: 'Setup' })
  }
  const savedRows = savedTable.row_order.filter(bindingId => bindingById.has(bindingId))
  const rowIds = [...savedRows, ...inputBindings.map(binding => binding.id).filter(bindingId => !savedRows.includes(bindingId))]
  for (const bindingId of rowIds) {
    const binding = bindingById.get(bindingId)!
    rows.push({ id: `binding:${binding.id}`, kind: 'ingredient', bindingId: binding.id, ingredientLineageId: binding.ingredient_lineage_id, label: binding.ingredient_lineage_id })
  }
  if (!rows.length && order.length) rows.push({ id: 'setup:default', kind: 'setup', label: 'Flow' })

  const rowIndex = new Map(rows.map((row, index) => [row.id, index]))
  const labels = new Map(savedTable.labels.map(label => [label.action_id, label]))
  for (const action of method.actions) {
    const label = labels.get(action.id)
    if (!label) warnings.push({ code: 'missing_label', message: 'An operation is missing a compact table label.', blocking: true, entityId: action.id })
    if (label && label.confidence < threshold && label.accepted !== true) warnings.push({ code: 'low_confidence_label', message: 'An operation label is uncertain.', blocking: true, entityId: action.id })
  }
  for (const binding of inputBindings) {
    if (binding.confidence < threshold && binding.accepted !== true) warnings.push({ code: 'low_confidence_binding', message: 'An ingredient placement is uncertain.', blocking: true, entityId: binding.id })
  }
  for (const edge of method.edges) {
    if (edge.confidence < threshold && edge.accepted !== true) warnings.push({ code: 'low_confidence_edge', message: 'A branch connection is uncertain.', blocking: true, entityId: edge.id })
  }

  const cellsByAction = new Map<string, FlowCell>()
  const logicalColumns = new Map<string, number>()
  const occupied = new Map<number, FlowCell[]>()
  const hints = new Map(savedTable.column_hints.map(hint => [hint.action_id, hint.preferred_column]))
  const rowRangeForAction = (actionId: string) => {
    const directRows = inputBindings
      .filter(binding => binding.action_id === actionId)
      .map(binding => rowIndex.get(`binding:${binding.id}`))
      .filter((index): index is number => index != null)
    const predecessorRows = (incoming.get(actionId) ?? [])
      .map(parentId => cellsByAction.get(parentId))
      .filter((cell): cell is FlowCell => Boolean(cell))
      .flatMap(cell => Array.from({ length: cell.rowSpan }, (_, offset) => cell.rowStart + offset))
    const values = [...directRows, ...predecessorRows]
    if (!values.length) {
      const setup = rowIndex.get(`setup:${actionId}`)
      return { rowStart: setup ?? 0, rowSpan: 1 }
    }
    const rowStart = Math.min(...values)
    return { rowStart, rowSpan: Math.max(...values) - rowStart + 1 }
  }
  for (const actionId of order) {
    const action = actions.get(actionId)!
    const parents = incoming.get(actionId) ?? []
    const earliest = parents.length ? Math.max(...parents.map(parentId => (logicalColumns.get(parentId) ?? 0) + 1)) : 0
    let logicalColumn = Math.max(earliest, hints.get(actionId) ?? 0)
    const range = rowRangeForAction(actionId)
    while ((occupied.get(logicalColumn) ?? []).some(cell => overlap(cell, range))) logicalColumn += 1
    const cell: FlowCell = {
      actionId,
      column: logicalColumn,
      rowStart: range.rowStart,
      rowSpan: range.rowSpan,
      inputBindingIds: inputBindings.filter(binding => binding.action_id === actionId).map(binding => binding.id),
      predecessorActionIds: parents,
      kind: parents.length > 1 ? 'merge' : setupIds.has(actionId) ? 'setup' : 'operation',
    }
    cellsByAction.set(actionId, cell)
    logicalColumns.set(actionId, logicalColumn)
    occupied.set(logicalColumn, [...(occupied.get(logicalColumn) ?? []), cell])
  }
  const logicalColumnValues = [...new Set([...logicalColumns.values()])].sort((a, b) => a - b)
  const columnIndex = new Map(logicalColumnValues.map((value, index) => [value, index]))
  const labelsByAction = new Map(savedTable.labels.map(label => [label.action_id, label.text]))
  const columns: FlowColumn[] = logicalColumnValues.map((logical, index) => {
    const actionId = order.find(id => logicalColumns.get(id) === logical)!
    const action = actions.get(actionId)!
    const stage = method.stages.find(item => item.id === action.stage_id)
    return { actionId, stageId: action.stage_id, stageTitle: stage?.title ?? 'Method', index, label: labelsByAction.get(actionId) ?? action.text }
  })
  const cells = [...cellsByAction.values()].map(cell => ({ ...cell, column: columnIndex.get(cell.column) ?? cell.column }))
  const connectors: FlowConnector[] = method.edges.flatMap(edge => {
    const from = cellsByAction.get(edge.from_action_id)
    const to = cellsByAction.get(edge.to_action_id)
    if (!from || !to) return []
    const fanIn = (incoming.get(edge.to_action_id) ?? []).length > 1
    const fanOut = (outgoing.get(edge.from_action_id) ?? []).length > 1
    return [{
      edgeId: edge.id,
      kind: fanIn ? 'merge' : fanOut ? 'fork' : edge.kind,
      fromActionId: edge.from_action_id,
      toActionId: edge.to_action_id,
      fromRow: from.rowStart,
      toRow: to.rowStart,
      column: Math.max(0, (columnIndex.get(to.column) ?? to.column) - 1),
    }]
  })
  const lanes: FlowLane[] = []
  for (const column of columns) {
    const last = lanes[lanes.length - 1]
    if (last?.stageId === column.stageId) last.endColumn = column.index
    else lanes.push({ stageId: column.stageId, title: column.stageTitle, startColumn: column.index, endColumn: column.index })
  }
  return { rows, columns, cells, connectors, lanes, warnings }
}

File: backend/app/services/recipe_flow_layout.py
"""Deterministic Flow table layout shared with the frontend contract tests."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..schemas import MethodDocument, MethodTableDocument


@dataclass(frozen=True)
class _Cell:
    action_id: str
    column: int
    row_start: int
    row_span: int
    input_binding_ids: tuple[str, ...]
    predecessor_action_ids: tuple[str, ...]
    kind: str


def _action_key(action: Any) -> tuple[int, str]:
    return int(action.position), action.id


def _overlaps(left: _Cell | dict[str, Any], right: dict[str, int]) -> bool:
    left_start = left.row_start if isinstance(left, _Cell) else left["row_start"]
    left_span = left.row_span if isinstance(left, _Cell) else left["row_span"]
    return left_start < right["row_start"] + right["row_span"] and right["row_start"] < left_start + left_span


def layout_flow_table(
    method: MethodDocument | dict[str, Any],
    table: MethodTableDocument | dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Return the same structural layout shape used by ``recipeFlowLayout.ts``."""

    document = method if isinstance(method, MethodDocument) else MethodDocument.model_validate(method)
    table_document = (
        table
        if isinstance(table, MethodTableDocument)
        else MethodTableDocument.model_validate(table)
        if table is not None
        else MethodTableDocument()
    )
    actions = {action.id: action for action in document.actions}
    incoming: dict[str, list[str]] = {action.id: [] for action in document.actions}
    outgoing: dict[str, list[str]] = {action.id: [] for action in document.actions}
    warnings: list[dict[str, Any]] = []
    for edge in document.edges:
        if edge.from_action_id not in actions or edge.to_action_id not in actions:
            warnings.append({
                "code": "dangling_edge",
                "message": "A flow connector points to an operation that no longer exists.",
                "blocking": True,
                "entityId": edge.id,
            })
            continue
        incoming[edge.to_action_id].append(edge.from_action_id)
        outgoing[edge.from_action_id].append(edge.to_action_id)
    indegree = {action_id: len(parents) for action_id, parents in incoming.items()}
    queue = sorted((action for action in document.actions if indegree[action.id] == 0), key=_action_key)
    order: list[str] = []
    while queue:
        action = queue.pop(0)
        order.append(action.id)
        for child_id in sorted(outgoing[action.id]):
            indegree[child_id] -= 1
            if indegree[child_id] == 0:
                queue.append(actions[child_id])
                queue.sort(key=_action_key)
    if len(order) != len(document.actions):
        warnings.append({
            "code": "graph_cycle",
            "message": "This flow contains a cycle; connectors are shown in source order until it is repaired.",
            "blocking": True,
        })
        order.extend(
            action.id
            for action in sorted(document.actions, key=_action_key)
            if action.id not in order
        )

    input_bindings = [binding for binding in document.ingredient_bindings if binding.role != "reference"]
    binding_by_id = {binding.id: binding for binding in input_bindings}
    setup_ids = {action_id for action_id in table_document.setup_action_ids if action_id in actions}
    rows: list[dict[str, Any]] = []
    for action_id in order:
        if action_id in setup_ids:
            rows.append({"id": f"setup:{action_id}", "kind": "setup", "actionId": action_id, "label": "Setup"})
    saved_rows = [binding_id for binding_id in table_document.row_order if binding_id in binding_by_id]
    row_ids = saved_rows + [binding.id for binding in input_bindings if binding.id not in saved_rows]
    for binding_id in row_ids:
        binding = binding_by_id[binding_id]
        rows.append({
            "id": f"binding:{binding.id}",
            "kind": "ingredient",
            "bindingId": binding.id,
            "ingredientLineageId": binding.ingredient_lineage_id,
            "label": binding.ingredient_lineage_id,
        })
    if not rows and order:
        rows.append({"id": "setup:default", "kind": "setup", "label": "Flow"})
    row_index = {row["id"]: index for index, row in enumerate(rows)}
    labels = {label.action_id: label for label in table_document.labels}
    for action in document.actions:
        label = labels.get(action.id)
        if label is None:
            warnings.append({"code": "missing_label", "message": "An operation is missing a compact table label.", "blocking": True, "entityId": action.id})
        elif label.confidence < 0.65 and not label.accepted:
            warnings.append({"code": "low_confidence_label", "message": "An operation label is uncertain.", "blocking": True, "entityId": action.id})
    for binding in input_bindings:
        if binding.confidence < 0.65 and not binding.accepted:
            warnings.append({"code": "low_confidence_binding", "message": "An ingredient placement is uncertain.", "blocking": True, "entityId": binding.id})
    for edge in document.edges:
        if edge.confidence < 0.65 and not edge.accepted:
            warnings.append({"code": "low_confidence_edge", "message": "A branch connection is uncertain.", "blocking": True, "entityId": edge.id})

    cells_by_action: dict[str, _Cell] = {}
    logical_columns: dict[str, int] = {}
    occupied: dict[int, list[_Cell]] = {}

    def row_range(action_id: str) -> tuple[int, int]:
        direct_rows = [
            row_index[f"binding:{binding.id}"]
            for binding in input_bindings
            if binding.action_id == action_id and f"binding:{binding.id}" in row_index
        ]
        predecessor_rows: list[int] = []
        for parent_id in incoming[action_id]:
            parent = cells_by_action.get(parent_id)
            if parent is not None:
                predecessor_rows.extend(range(parent.row_start, parent.row_start + parent.row_span))
        values = direct_rows + predecessor_rows
        if not values:
            return row_index.get(f"setup:{action_id}", 0), 1
        start = min(values)
        return start, max(values) - start + 1

    hints = {hint.action_id: hint.preferred_column for hint in table_document.column_hints}
    for action_id in order:
        parents = incoming[action_id]
        earliest = max((logical_columns[parent_id] + 1 for parent_id in parents), default=0)
        column = max(earliest, hints.get(action_id, 0))
        start, span = row_range(action_id)
        while any(_overlaps(cell, {"row_start": start, "row_span": span}) for cell in occupied.get(column, [])):
            column += 1
        cell = _Cell(
            action_id=action_id,
            column=column,
            row_start=start,
            row_span=span,
            input_binding_ids=tuple(binding.id for binding in input_bindings if binding.action_id == action_id),
            predecessor_action_ids=tuple(parents),
            kind="merge" if len(parents) > 1 else "setup" if action_id in setup_ids else "operation",
        )
        cells_by_action[action_id] = cell
        logical_columns[action_id] = column
        occupied.setdefault(column, []).append(cell)

    logical_values = sorted(set(logical_columns.values()))
    column_index = {value: index for index, value in enumerate(logical_values)}
    labels_by_action = {label.action_id: label.text for label in table_document.labels}
    columns: list[dict[str, Any]] = []
    for index, logical in enumerate(logical_values):
        action_id = next(action_id for action_id in order if logical_columns[action_id] == logical)
        action = actions[action_id]
        stage = next((stage for stage in document.stages if stage.id == action.stage_id), None)
        columns.append({
            "actionId": action_id,
            "stageId": action.stage_id,
            "stageTitle": stage.title if stage is not None else "Method",
            "index": index,
            "label": labels_by_action.get(action_id, action.text),
        })
    cells = [
        {
            "actionId": cell.action_id,
            "column": column_index[cell.column],
            "rowStart": cell.row_start,
            "rowSpan": cell.row_span,
            "inputBindingIds": list(cell.input_binding_ids),
            "predecessorActionIds": list(cell.predecessor_action_ids),
            "kind": cell.kind,
        }
        for cell in cells_by_action.values()
    ]
    connectors: list[dict[str, Any]] = []
    for edge in document.edges:
        source = cells_by_action.get(edge.from_action_id)
        target = cells_by_action.get(edge.to_action_id)
        if source is None or target is None:
            continue
        fan_in = len(incoming[edge.to_action_id]) > 1
        fan_out = len(outgoing[edge.from_action_id]) > 1
        connectors.append({
            "edgeId": edge.id,
            "kind": "merge" if fan_in else "fork" if fan_out else edge.kind,
            "fromActionId": edge.from_action_id,
            "toActionId": edge.to_action_id,
            "fromRow": source.row_start,
            "toRow": target.row_start,
            "column": max(0, column_index[target.column] - 1),
        })
    lanes: list[dict[str, Any]] = []
    for column in columns:
        previous = lanes[-1] if lanes else None
        if previous is not None and previous["stageId"] == column["stageId"]:
            previous["endColumn"] = column["index"]
        else:
            lanes.append({
                "stageId": column["stageId"],
                "title": column["stageTitle"],
                "startColumn": column["index"],
                "endColumn": column["index"],
            })
    return {
        "rows": rows,
        "columns": columns,
        "cells": cells,
        "connectors": connectors,
        "lanes": lanes,
        "warnings": warnings,
    }


create_flow_table_layout = layout_flow_table
